### ISSUE TITLE
Regression: custom batches no longer supported

### ISSUE THREAD

**Issue by AmitMY** [joeynmt/joeynmt#188]
**Describe the bug**
In `joeynmt<2.0.0`, the `TrainManager` used to receive a `batch_class` to handle custom batches.
In `joeynmt=2.0.0`, it is no longer needed, as there is the `make_data_iter` which calls a `collate_fn` responsible to return whatever.

This created a regression - it used to be possible to use this repository as a library - but now one has to make hard-coded changes in the code in order to have a custom `make_data_iter`.

It is hard coded in [training](https://github.com/joeynmt/joeynmt/blob/main/joeynmt/training.py#L390) and [prediction](https://github.com/joeynmt/joeynmt/blob/main/joeynmt/prediction.py#L112)

**Possible Solutions:**
For training:
1. to have `data_iter` callable passed to the `TrainManager`
2. in the train manager, have a method so we can extend it:
```py
def make_data_iter(self, **kwargs):
  return make_data_iter(**kwargs)
```

For prediction:
1. pass this as an argument to the `predict` method?

----

What solutions would you prefer/approve if any?

**Comment by may-** [comment]
Hi @AmitMY,

Do you mean you want to have your own custom batch class?
The main point of the refactoring v1 -> v2 was to get rid of the torchtext dependency and to have an own `collate_fn()`. How about overriding the `collate_fn()` for your custom batch class? I guess a different batch class implies a different data type, a different data structure, doesn't it? If so, I recommend to create a new dataset class which has a flag in which `batch_class` should be invoked.

First, create a new batch class in batch.py:
```python
class CustomBatch(Batch):
    def __init__():
        # whatever you want
```

Second, implement a new `collate_fn()` in data.py
```python
def custom_collate_fn(batch, ...):
    # whatever you want
    # Note: `batch` arg here is coming from `__getitem__()` func in the dataset object.
    # src, trg = zip(*batch)
    return CustomBatch(src=src, trg=trg, ...)

def make_data_iter(dataset, ...):
    if dataset.batch_class == "custom_batch":
        return DataLoader(dataset, collate_fn=custom_collate_fn, ...)
    else:  # default batch class
        return DataLoader(dataset, collate_fn=collate_fn, ...)
```

Third, create a new dataset class in datasets.py:
```python
class CustomDataset(PlainDataset):
    def __init__(self, path, ...):
        # whatever you want
        self.data = read_data_from_file(path)
        self.batch_class = kwargs.get("batch_class", "default")

    def __getitem__(self, index: int):
        # what you return here appears in `batch` argument in `collate_fn()`.
        # For example, if you return a tuple of two lists such as 
        # `(['a', 'b'], ['c', 'd', 'e'])` for a datapoint of index = 1 and
        # `(['x', 'y', 'z'], ['v', 'w'])` for a datapoint of index = 2, and assume batch_size = 2,
        # then you will get a batch of these two datapoints as a list:
        # `[(['a', 'b'], ['c', 'd', 'e']), (['x', 'y', 'z'], ['v', 'w'])]` 
        # in your `batch` arg in `collate_fn()`
        x, y = self.data[index]
        return x, y

def build_dataset(dataset_type, ...):
    if dataset_type == "plain":
        dataset = PlainDataset(path=path, ...)
        ...
    elif dataset_type == "custom_dataset":
        dataset = CustomDataset(path=path, ...)
```

Then you can choose your custom dataset in the config file:
```yaml
data:
    train: "path/to/train"
    dev: "path/to/dev" 
    test: "path/to/test"
    dataset_type: "custom_dataset"
...
```
Do you think does it make sense for your use case?

I personally feel it's not so appropriate to change `TrainManager` according to batch class, because a batch type is rather related to datatype, not (directly) to the training algorithm. Or do you think `TrainManager` is the conceptually correct place to resolve this issue? We are open for any further suggestions!

**Comment by may-** [comment]
@juliakreutzer, what do you think this issue?
Do you want to put `data_class` variable as an argument of `TrainManager` back? It was more or less intentional to remove this argument from `TrainManager` in version 2.

**Comment by AmitMY** [comment]
Hi @may- maybe I wasn't clear -
The functionality I am looking for is using `joey` *as a library*, meaning, not modifying the code, but mostly importing `joeynmt` and using it.

This was possible before, by using a `batch_class` to have a custom type of data - [added by me](https://github.com/joeynmt/joeynmt/pull/124/) - as in my example I had two sequences as input, and needed a custom data object.

Now while it is possible to define my own `collate_fn`, and I did so in `2.0.0`, I can not "override" it anywhere. If I want to override it, I need to clone ALL of joey, and replace the imports in the training and prediction scripts.

So, the regression is only in importing joey as is, not in changing it, because it is more easy to maintain that way.

For reference, here is my code: https://github.com/sign-language-processing/transcription/tree/joey/pose_to_text (you will see that the files are short because i'm mostly importing, not overriding)

**Comment by may-** [comment]
Hi @AmitMY,

humm, ok, so the question is whether to merge it into the official joey or not, right? Honestly, I'm not sure if it is necessary for the official joey. You can always keep your importable version of joey with a custom `batch_class` in your local, and you can import your version of joey, can't you?

I'd like to wait for @juliakreutzer's decision. Maybe we can call for votes from other contributors democratically??

**Comment by AmitMY** [comment]
I have tried to go a different way - and adapt whatever I could to the existing code, but eventually got stuck with a single unfortunate issue:
https://github.com/joeynmt/joeynmt/blob/dc854b228f39af83b23035ce01f614ee0b072c70/joeynmt/data.py#L201

My data is continuous, so even though I managed to override the tokenizer, and create a custom `sequence_encoder` method to collate that continuous data, this `.long()` casting breaks.

Having the ability to *only* modify `collate_fn` to one's needs would be awesome, but I understand that overriding `make_data_iter` would be easier.

Using `joeynmt` as a library guarantees more code readability for users, less room for errors, more tight control on what the user can and cannot modify, and creates lower code footprints that are more maintainable.

**Comment by juliakreutzer** [comment]
Hi @AmitMY and @may- 

thanks for raising this point @AmitMY , and thank you @may- for suggesting solutions and considering other users as well.

I agree with the point that it would be nice if joeynmt could be used as a library as well. Like @may- said, it's not the primary purpose it was designed for, so we don't have tests that reflect that, and we have to be considerate when making API changes, especially those introducing more non-obvious abstractions. 

In order to prevent the discussion from getting too general (there might be other changes needed if other uses wanted to use it as a library as well, with other customizations and overrides), I suggest we focus on the minimal changes that would be needed to let @AmitMY work with his code as before.

I agree with @may- that the TrainManager would be cleaner if it didn't receive data-related inputs. However, train_and_validate is also tied to it, which requires data, obviously. So perhaps a solution that goes into the direction of having custom data iterators (@AmitMY's suggestion number 2) would be best?
How about pulling the data iterator creation (`make_data_iter`) out of `train_and_validate` (https://github.com/joeynmt/joeynmt/blob/main/joeynmt/training.py#L382) and `predict` (https://github.com/joeynmt/joeynmt/blob/6c580f838ad19ea22222a8ea14ca0a371533aef4/joeynmt/prediction.py#L41) and instead pass data iterators directly to both functions (rather than datasets), or a `make_data_iter` function as argument that can be overridden? I think that would also make the code cleaner, because the data iterator creator could happen in a more central place (outside train or predict). 
You two are way deeper into the code than me :smile: @AmitMY would that be sufficient for your use case? @may- what do you think?

**Comment by may-** [comment]
Hi @juliakreutzer,

thank you for your suggestion.

When I removed the torchtext's datadoader, I intentionally omit passing the custom batch class as an argument. Because it will give a wrong impression that joey can be used for different data type (like continuous data) without implementing collate function and just overriding the batch class. But if the data type is different, I strongly recommend to implement their own collate function (padding function) for the data type. I don't want to enable instant overriding the batch class without changing adequate padding.

I'm still skeptical how much beneficial to merge this into the main repo. I personally believe that passing the custom batch class as an argument makes the code less readable and more difficult to extend. When I extended joey for speech input (continuous data), the custom batch class as an argument was rather confusing. To me, it's kind of a back step to the torchtext-style dependency. It looks at the first glance more flexible but hide the necessary steps of data loading, and therefore less meaningful for educational purpose. If there are uses who need this functionality, couldn't we keep that modified version in a different repo or branch as a spin-off?

**Comment by AmitMY** [comment]
Hi @may- and @juliakreutzer 
I have created the minimal changes I feel are necessary to allow this customization.
Please let me know what you think.

**Comment by may-** [comment]
@AmitMY 

now I understand better what you meant.
yeah, it's indeed nice to have, I agree.
sorry, if my words sound too harsh...

**Comment by AmitMY** [comment]
Don't worry about it, I could have probably made myself clearer.

PR is ready to be reviewed. 
@juliakreutzer could you give it a look? I'd recommend reviewing it one-commit-at-a-time as each has a specific change. (https://github.com/joeynmt/joeynmt/pull/189/commits/7a03d297373bdcb8a154e42955ab08ed0660e119 then https://github.com/joeynmt/joeynmt/pull/189/commits/0fb08b8d5e276d026528c2c36c49217618119854)

**Comment by may-** [comment]
@AmitMY 

I feel now, we can merge all functions in `data.py` into `data_loader.py`? How do you think about this?
`make_data_iter()` is literaly a member of the data_loader class, and all others (samplers) belongs to `make_data_iter()`, actually.

**Comment by AmitMY** [comment]
Yeah I understand three files might be too many, but it's the same amount of code - 
the realistic options are:
- 3 files
- move the one function in `data_loader.py` to `datasets.py`, delete `data_loader.py`
- move *all* 200+ lines from `data.py` to `datasets.py`, rename `data_loader.py` to `data.py`

Because `make_data_iter` depends on everything in `data.py`, but `load_data` only depends on `build_dataset`


What solution do you prefer? Would you prefer solving this circular import yourself?

**Comment by may-** [comment]
I prefer the 3rd one: move all 200+ lines from `data.py` to `datasets.py`, rename `data_loader.py` to `data.py`. Is `datasets.py` too long, then?

you are right, it's the same amount of code. do you think the 3 files version would be easier to understand the structure for new comers??

I don' like my dirty hack with TYPE_CHECKING to avoid the circular import, your solutions (separating files or moving functions into `datasets.py`) are way better for newbie users 😃

**Comment by AmitMY** [comment]
I feel like the 3-file solution is more easy to understand, if only just because it doesn't create one massive `datasets` file

but i'm happy to go for the 2-file solution if that makes it so it would be approved. just say the word

**Comment by AmitMY** [comment]
@may- I have changed things to be like the 3rd option. Please let me know if anything else needs changing in your opinion

**Comment by may-** [comment]
@AmitMY thank you for your contribution! looks great. If other (newbie) users also think the 3 file version is easier to follow, put the structure back then.

**Comment by AmitMY** [comment]
I moved it to two files as @may- and @juliakreutzer both thought "three classes for data-related functions might be confusing".

I'd argue for merging and releasing as is (to support the functionality), and if down the line some restructuring needs to be done, that's fine.
